// !RENDER_DIAGNOSTICS_FULL_TEXT
// FIR_DUMP

fun foo() {
    buildList {
        <!RECEIVER_TYPE_MISMATCH_WARNING_FOR_BUILDER_INFERENCE!>add<!>("Boom")
        <!RECEIVER_TYPE_MISMATCH_WARNING_FOR_BUILDER_INFERENCE!>println<!>(<!RECEIVER_TYPE_MISMATCH_WARNING_FOR_BUILDER_INFERENCE!><!RECEIVER_TYPE_MISMATCH_WARNING_FOR_BUILDER_INFERENCE!>plus<!>(1)[0]<!>)
    }
}

